src/sentry/rules/actions/utils.py

```diff
@@ -31,41 +31,38 @@ def check_value_changed(
 
 
 def generate_diff_labels(
-    present_state: dict[str, Any],
-    prior_state: dict[str, Any],
+    present_state: list[dict[str, Any]],
+    prior_state: list[dict[str, Any]],
     rule: Rule,
-    changed_data: DefaultDict[str, list[str]],
     key: str,
 ) -> DefaultDict[str, list[str]]:
     added_statement = "Added {} '{}'"
     removed_statement = "Removed {} '{}'"
     added_items = []
-
+    changed_data: DefaultDict[str, list[str]] = defaultdict(list)
     # Added items that are not seen in the prior rule data
-    for data in prior_state.get(key, []):
-        if data not in present_state.get(key, []):
+    for data in prior_state:
+        if data not in present_state:
             label = generate_rule_label(rule.project, rule, data)
-            changed_data[data["id"]].append(added_statement.format(key[:-1], label))
+            changed_data[data["id"]].append(added_statement.format(key, label))
             added_items.append(data["id"])
 
     # Check if the id is in both but the data has changed
     for item in added_items:
         if changed_data.get(item):
-            for data in prior_state.get(key, []):
-                for present_data in present_state.get(key, []):
-                    if data["id"] == item and present_data["id"] == item:
-                        if present_data != data:
-                            old_label = generate_rule_label(rule.project, rule, data)
-                            new_label = generate_rule_label(rule.project, rule, present_data)
-                            changed_data[item] = [
-                                (f"Changed {key[:-1]} from *{old_label}* to *{new_label}*")
-                            ]
+            prior_item = [data for data in prior_state if data["id"] == item]
```

Comment:
`prior_items`?

Also - can we precompute this beforehand too?
construct like a `items_by_id` map and pass the map into this function?
same with present items
this way we don't need to do so many loops within loops.

this can just become 
```
prior_items = prior_items_by_id.get(id)
```

Reply:
🤔 I'm not sure if that would work, it's passed in a list of dictionaries that are guaranteed to have an `id`  key, but the rest of the data is not guaranteed - it could be nothing (`{'id': 'sentry.rules.conditions.first_seen_event.FirstSeenEventCondition'}` for example) or it could be several items (`{'channel_id': 'new_channel_id', 'workspace': '1', 'id': 'sentry.integrations.slack.notify_action.SlackNotifyServiceAction', 'channel': 'new_channel_name', 'tags': '', 'uuid': '5cefca58-ecbb-47e4-98b1-68aef5cc4c7f'}` for example).

Reply:
I updated it to what I _think_ you were saying, but it's a little funny because `generate_rule_label` needs to have the id as well, so they look like:

`{'sentry.rules.conditions.first_seen_event.FirstSeenEventCondition': [{'id': 'sentry.rules.conditions.first_seen_event.FirstSeenEventCondition'}]}`

`{'sentry.rules.conditions.first_seen_event.FirstSeenEventCondition': [{'id': 'sentry.rules.conditions.first_seen_event.FirstSeenEventCondition'}], 'sentry.rules.filters.issue_category.IssueCategoryFilter': [{'id': 'sentry.rules.filters.issue_category.IssueCategoryFilter', 'value': '2'}]}`


`{'sentry.integrations.slack.notify_action.SlackNotifyServiceAction': [{'channel_id': 'old_channel_id', 'workspace': '1', 'id': 'sentry.integrations.slack.notify_action.SlackNotifyServiceAction', 'channel': '#old_channel_name'}]}`

`{'sentry.integrations.slack.notify_action.SlackNotifyServiceAction': [{'channel_id': 'new_channel_id', 'workspace': '1', 'id': 'sentry.integrations.slack.notify_action.SlackNotifyServiceAction', 'channel': 'new_channel_name', 'tags': '', 'uuid': '667d60c8-b7a0-4412-a491-c99ba8e8ec0b'}]}`

Reply:
This seems fine to me?

the contents of the data don't matter.
The value here is that we don't need to go through the whole rigamarole of N^2 compilation